app/src/main/java/org/greenstand/android/TreeTracker/managers/Locations.kt

```diff
@@ -0,0 +1,332 @@
+                        longitude,
+                        accuracy,
+                        generatedTreeUuid?.toString() ?: null,
+                        System.currentTimeMillis()
+                    )
+                val jsonValue = gson.toJson(locationData)
+                Timber.d("Inserting new location data $jsonValue")
+                treeTrackerDAO.insertLocationData(LocationDataEntity(jsonValue))
+            }
+
+            if (isInTreeCaptureMode() && !convergenceWithinRange) {
+
+                var evictedLocation: Location? = if (locationsDeque.size >= CONVERGENCE_DATA_SIZE)
+                    locationsDeque.pollFirst() else null
+                locationsDeque.add(it)
+
+                if (locationsDeque.size >= CONVERGENCE_DATA_SIZE) {
+                    if (convergence == null) {
+                        convergence = Convergence(locationsDeque.toList())
+                        convergence!!.computeConvergence()
+                    } else {
+                        convergence!!.computeSlidingWindowConvergence(evictedLocation!!, it)
+                    }
+                    Timber.d(
+                        "Convergence: Longitude Mean: " +
+                                "[${convergence!!.longitudeConvergence?.mean}]. \n" +
+                                "Longitude standard deviation value: " +
+                                "[${convergence!!.longitudeConvergence?.standardDeviation}]"
+                    )
+                    Timber.d(
+                        "Convergence: Latitude Mean: " +
+                                "[${convergence!!.latitudeConvergence?.mean}]. \n " +
+                                "Latitude standard deviation value: " +
+                                "[${convergence!!.latitudeConvergence?.standardDeviation}]"
+                    )
+                    safeLet(
+                        convergence!!.longitudinalStandardDeviation(),
+                        convergence!!.latitudinalStandardDeviation()
+                    ) { longStdDev, latStdDev ->
+                        convergenceWithinRange = longStdDev < LONG_STD_DEV &&
+                                latStdDev < LAT_STD_DEV
+                    }
+                }
+            }
+        }
+    }
+
+    fun start() {
+        locationUpdateManager.locationUpdateLiveData.observeForever(locationObserver)
+    }
+
+    fun isInTreeCaptureMode(): Boolean {
+        return generatedTreeUuid != null
+    }
+
+    fun turnOnTreeCaptureMode() {
+        generatedTreeUuid = UUID.randomUUID()
+        Timber.d("Convergence: Tree capture mode turned on")
+    }
+
+    fun turnOffTreeCaptureMode() {
+        generatedTreeUuid = null
+        convergence = null
+        locationsDeque.clear()
+        convergenceWithinRange = false
+        Timber.d("Convergence: Tree capture turned off")
+    }
+}
+
+class Convergence(val locations: List<Location>) {
+
+    var longitudeConvergence: ConvergenceStats? = null
```

Comment:
I'm curious if it would help if there was an invalid ConvergenceStats state that this is initially set to instead of null. That way we can avoid all the double bands and null checks.

Reply:
Hmm, I see having null as appropriate. The only usage for these variables are behind a safe `?` operator. So we are good.